### PR TITLE
update test/perf/micro from 0.4.0. to 0.7.0

### PR DESCRIPTION
**Pull request by johnfgibson**
The micro benchmarks shown on julialang.org and julialang.org/benchmarks are out-of-date: the julia data is for julia-0.4.0, and other languages are comparably old. This set of changes gets test/perf/micro running again for julia-0.7.0-DEV so that fresh benchmarks can be run and new results published. 

There are still a few outstanding questions on the [fresh benchmark results](https://discourse.julialang.org/t/benchmark-for-latest-julia/5991/22)  (several languages including Fortran are surprisingly slow on `rand_mat_mul` and several suprisingly fast on `mandel`). But I'd like to get PR merged so that others can run the benchmarks and help dig into their behavior. 

Changes in test/perf/Makefile reflect revised locations for openblas and other libs in julia source tree sometime since 0.4.0. Commented out Scala and Go benchmarks since I couldn't get these systems installed and running on my local machine and their failure broke "make benchmarks.csv".

Switched from gsl-shell lua to scilua, per @stepelu's recommendation for issue #14222.

Minor Matlab/octave syntax update in perf.m required for octave-4.0.3

### CODE REVIEW

**Comment by johnfgibson** [comment]
This last set of revisions expands the benchmark names to emphasize the algorithm over the computation (e.g. fib -> recursion_fibonacci), reorders languages in plots and tables to C, Julia, LuaJIT, ... following geometric mean after Julia, and incorporates BLAS calls wherever possible. 

There a corresponding pull request [#643 for julialang.github.com](https://github.com/JuliaLang/julialang.github.com/pull/643) to update the benchmark table and plot on the julialang.org website.

**Comment by StefanKarpinski** [comment]
Nevermind: I see you already changed it to LuaJIT. Thanks so much for putting in the work to do this. Getting all of those benchmarks running is very much non-trivial, so it's greatly appreciated and we were sorely in need of an update!

**Comment by johnfgibson** [comment]
My first pull request merged into the Julia source tree! Can I just say I am thrilled?!?!  I am so glad to to be able to contribute to this fantastic project.

**Comment by StefanKarpinski** [comment]
I'm thrilled too! Love the new plot.

**Comment by KristofferC** [comment]
Very nice job setting up the whole stack to run all the benchmarks! I also like the renaming since it will cause people to write less blogs how they can beat Julia's fibonnacci benchmark by using a lookup table...

**Comment by pkofod** [comment]
> My first pull request merged into the Julia source tree!

You might want to visit your doctor, 'cus you've just been infected my friend! Congrats.

File: test/perf/micro/perf.f90 [hunk 1]
```diff
@@ -190,6 +190,15 @@ integer function parse_int(s, base) result(n)
 
 end function
 
+subroutine printfd(n)
+integer, intent(in) :: n
+integer :: i , unit
+open(unit=1, file="foo")
```
Comment (certik) [inline test/perf/micro/perf.f90]: Don't use units directly like that. Rather use `newunit`:
```fortran
integer :: u
open(newunit=u, file="foo", status="replace")
```
Also, very important: you must put the `status` in there. Per the Fortran standard, if the `status` specifier is omitted, the default value is `UNKNOWN`. The `UNKNOWN` status is processor dependent.

See here for more details: http://www.fortran90.org/src/best-practices.html#file-input-output

File: test/perf/micro/perf.f90 [hunk 2]
```diff
@@ -190,6 +190,15 @@ integer function parse_int(s, base) result(n)
...
+do i = 1, n
+  write(unit=1, fmt=*) i, i
+end do
+close(unit=1)
```
Comment (certik) [inline test/perf/micro/perf.f90]: So this subroutine should be written as follows:
```fortran
subroutine printfd(n)
integer, intent(in) :: n
integer :: i, u
open(newunit=u, file="foo", status="replace")
do i = 1, n
  write(u, *) i, i
end do
close(u)
end subroutine
```

**Review by certik — COMMENTED** [review]
I noticed a few things that should be fixed in the Fortran code.

File: test/perf/micro/perf.f90 (hunk 2 above)
Comment (johnfgibson) [reply test/perf/micro/perf.f90]: @certik: thanks, I'll merge those fixes in.

**Pull request merged**